Support correlated scalar subqueries (v6)

kokosing opened the pull request:
Support correlated count scalar subqueries

A scalar subquery is a simple query which aggregates all the rows into a
single group (empty GROUP BY clause). For example:

SELECT min(a) FROM t WHERE b > 0
This commit add support for correlated count scalar subqueries which
have WHERE cluase of conjunction of equality comparison. For example:

SELECT (SELECT min(a) FROM t1(a,c) WHERE b = 0 AND c = d) FROM t2(b,d)
Logically such query is rewritten into left outer join where correlation
used in eqality comparisons (from WHERE cluase) are used as join
condition.

kokosing commented:
Supersedes #6035

martint commented on presto-main/src/main/java/com/facebook/presto/operator/AssignUniqueIdOperator.java:
```diff
@@ -32,8 +32,8 @@
 {
-    private static final long ROW_IDS_PER_REQUEST = 1 << 20;
-    private static final long MAX_ROW_ID = 1 << 40;
+    private static final long ROW_IDS_PER_REQUEST = 1L << 20L;
```
This should go in the commit that introduced the new node

kokosing replied:
done

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java:
```diff
@@ -0,0 +1,421 @@
+ * It always returns single row in Presto.
+ * <p>
+ * This optimizer can rewrite correlated scalar aggregation subquery to left outer join in a way described here:
+ * https://docs.google.com/document/d/18HN7peS2eR8lZsErqcmnoWyMEPb6p4OQeidH1JP_EkA/edit#heading=h.hb6ihepq7q74
```
This should link to the wiki document

kokosing replied:
done

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformCorrelatedScalarAggregationToJoin.java:
```diff
@@ -0,0 +1,421 @@
+ * From:
+ * - Apply (with correlation list: [C])
+ *   - (input) plan which produces symbols: [A, B, C]
+ *   - (subquery) Aggregation(GROUP BY 1; functions: [sum(F), count(), ...]
```
This should say GROUP BY ()

kokosing replied:
done

kokosing commented:
@martint 
I fixed the commit message. Previously I have two commits one for `count` and second for other aggregate functions. Then I merged them, somehow I failed to merged commit messages. Sorry about that.

The query you posted is somehow tricky. `EXISTS` subquery ends with double (nested) aggregation. Such query is not yet supported, though it unveiled the bug. `decorrelateFilters` could return still correlated node. I added a check (and tests) for that. 

I will try to extend the support for correlated subqueries to handle queries the one you posted.

kokosing commented:
@martint Rebased and all the comments addresssed. Superseded by #6190.
